src/hotspot/share/gc/g1/g1CollectedHeap.cpp

```diff
@@ -400,6 +402,48 @@ G1CollectedHeap::mem_allocate(size_t word_size,
   return attempt_allocation(word_size, word_size, &dummy);
 }
 
+bool G1CollectedHeap::attempt_allocation_after_gc(size_t word_size,
+                                                  uint gc_count_before,
+                                                  bool should_try_gc,
+                                                  HeapWord** result,
+                                                  GCCause::Cause gc_cause) {
+
+  StalledAllocReq request(word_size, _allocator->current_node_index());
+
+  {
+    MutexLocker ml(&_alloc_request_lock, Mutex::_no_safepoint_check_flag);
+    _stalled_allocations.insert_last(&request);
+  }
+
+  if (should_try_gc) {
+    do_collection_pause(word_size, gc_count_before, gc_cause);
+  } else {
+    log_trace(gc, alloc)("%s: Stall until clear", Thread::current()->name());
+    // The GCLocker is either active or the GCLocker initiated
+    // GC has not yet been performed. Stall until it is completed.
+    GCLocker::stall_until_clear();
+  }
+
+  {
+    MutexLocker ml(&_alloc_request_lock, Mutex::_no_safepoint_check_flag);
+    _stalled_allocations.remove(&request);
+  }
+
+  if (request.pending()) {
+    // GC Safepoint did not handle our allocation request. We should retry.
+    return false;
```

Comment:
Is it possible for a thread to hit this point indefinitely, i.e. some form of starvation?

Reply:
Only if you have indefinite triggering of non-allocating GCs. I wouldn't consider that starvation, as starvation here should imply some threads being able to allocate, and some other threads being unable to allocate even with allocation GCs triggered.